presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/Flatbush.java

```diff
@@ -0,0 +1,320 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A fast, low memory footprint static RTree.
+ * <p>
+ * Flatbushes create very few objects, instead storing the tree in a flat array.
+ * They support the standard RTree queries, but cannot be modified once built.
+ * It is quite possible to semi-efficiently remove objects.
+ * <p>
+ * Consider an RTree with branching factor `b`.  Each non-leaf node has two
+ * pieces of information:
+ * 1. The minimum bounding box of all descendants, and
+ * 2. Pointers to its children.
+ * <p>
+ * The former is simply four doubles.  The latter can be derived from the
+ * node's level and which sibling it is.  This means we can actually flatten
+ * the tree into a single array of doubles, four per node.  We can
+ * programmatically find the indices of a node's children (it will be faster if
+ * we pre-compute level offsets), and do envelope checks with just float
+ * operations.  "padded" empty nodes will have NaN entries, which will naturally
+ * return false for all comparison operators, thus being automatically not
+ * selected.
+ * <p>
+ * A critical choice in RTree implementation is how to group leaf nodes as
+ * children (and recursively, their parents).  One method that is very efficient
+ * to construct and comparable to best lookup performance is sorting by an
+ * object's Hilbert curve index.  Hilbert curves are naturally hierarchical, so
+ * successively grouping children and their parents will give a naturally nested
+ * structure.  This means we only need to sort the items once.
+ * <p>
+ * If sort time is a problem, we actually just need to "partition" into groups
+ * of `degree`, since the order of the children of a single parent doesn't
+ * matter.  This could be done with quicksort, stopping once all items at index
+ * `n * degree` are correctly placed.
+ * <p>
+ * Original implementation in Javascript: https://github.com/mourner/flatbush
+ */
+public class Flatbush<T extends HasExtent>
+{
+    private static final int FLATBUSH_INSTANCE_SIZE = ClassLayout.parseClass(Flatbush.class).instanceSize();
+    private static final int DEFAULT_DEGREE = 16;
+    // Number of coordinates to define an envelope
+    private static final int ENVELOPE_SIZE = 4;
+    // How many children per node
+    private final int degree;
+    // Offsets in tree for each level
+    private final int[] levelOffsets;
+    // Each node will have four doubles: xMin, yMin, xMax, yMax
+    private final DoubleBigArray tree;
+    private final T[] items;
+
+    public Flatbush(T[] items, int degree)
+    {
+        checkArgument(degree > 0, "degree must be positive");
+        this.degree = degree;
+        this.items = requireNonNull(items, "items is null");
+        this.levelOffsets = Flatbush.calculateLevelIndices(items.length, degree);
```

Comment:
- Flatbush.calculateLevelIndices -> calculateLevelIndices
- calculateLevelIndices -> calculateLevelOffsets to avoid using multiple terms for the same thing